src/main/java/com/aphyr/riemann/client/LocalhostResolver.java

```diff
@@ -0,0 +1,93 @@
+package com.aphyr.riemann.client;
+
+import java.net.UnknownHostException;
+
+/**
+ * A "Smarter" localhost resolver
+ * see issue: https://github.com/aphyr/riemann-java-client/issues/44
+ * Trying to avoid a lot of calls to java.net.InetAddress.getLocalHost()
+ * which under AWS trigger DNS resolving and have relatively high latency *per event*
+ * usually, the hostname doesn't change so often to warrant a real query.
+ *
+ * A real call to java.net.InetAddress.getLocalHost().getHostName()
+ * is made only if:
+ * 1) the refresh interval has passed (=result is stale)
+ * AND
+ * 2) no env vars that identify the hostname are found
+ */
+public class LocalhostResolver {
+
+    // default hostname env var names on Win/Nix
+    public static final String COMPUTERNAME = "COMPUTERNAME"; // Windows
+    public static final String HOSTNAME = "HOSTNAME"; // Nix
+
+    // how often should we refresh the cached hostname
+    public static long refreshIntervalMillis = 60 * 1000;
+
+    // cached hostname result
+    private static boolean envResolved = false;
+    private static String hostname;
+
+    // update (refresh) time management
+    private static long lastUpdate = 0;
+    public static long getLastUpdateTime() { return lastUpdate; }
+    // this is mostly for testing, plus ability to force a refresh
+    public static void setLastUpdateTime(long time) {
+        lastUpdate = time;
+    }
+
+    static {
+        resolveByEnv();
+    }
+
+    /**
+     * get resolved hostname.
+     * encapsulates all lookup and caching logic.
+     *
+     * @return the hostname
+     */
+    public static String getResolvedHostname() {
+        long now = System.currentTimeMillis();
+        if((now - refreshIntervalMillis) > lastUpdate) {
+            refreshResolve();
+        }
+
+        return hostname;
+    }
+
+    /**
+     * forces a new resolve even if refresh interval has not passed yet
+     */
+    public static void refreshResolve() {
```

Comment:
A few questions:
1. @aphyr Are you sure you actually want to support the case that the reverse DNS lookup for the local machine the code is running on has changed? How often does that happen ? Wouldn't it just confuse the riemann user more that the same stream now has a different hostname ? Aren't we better off, with performing the lookup only once?
2. @lassic Given we do want to refresh the hostname... why are we refreshing the hostname only if it is null or empty? According to aphyr's requirement on #45 the getLocalHost() should be called every time interval (assuming no env variables).
3. @aphyr - Isn't it advisable for getResolvedHostname() to have predictable low latency 100% of the calls? In that respect a separate thread is a good idea since the user's (low latency) thread is never used for performing a DNS call (with unpredictable latency). As for the teardown concern for the timer thread... doesn't client.scheduler() internals already take care of the teardown?